File: workshop/example/python/expression.py
from parser import Consecutive, OneOf, character, first, lazy, many1, satisfy, succeed

def expression():
    return first(OneOf([
        Consecutive([
            term(),
            expression_operator(),
            lazy(expression) 
        ]).map(lambda result: result[1](result[0], result[2])),
        term()
    ]))

def expression_operator():
    return first(OneOf([
        character('+').map(lambda _: (lambda l,r: Addition(l,r))),
        character('-').map(lambda _: (lambda l,r: Subtraction(l,r)))
    ]))

def term ():
    return first(OneOf([
        Consecutive([
            factor(),
            term_operator(),
            lazy(term) 
        ]).map(lambda result: result[1](result[0], result[2])),
        factor()
    ]))

def term_operator():
    return first(OneOf([
        character('*').map(lambda _: (lambda l,r: Multiplication(l,r))),
        character('/').map(lambda _: (lambda l,r: Division(l,r)))
    ]))

def factor():
    return OneOf([
        integer(),
        identifier(),
        parenthesized(lazy(expression))
    ])

def parenthesized(parser):
    return pack(character('('), parser, character(')'))

def pack(opening, parser, closing):
    return Consecutive([
        opening,
        parser,
        closing
    ]).map(lambda result: result[1])

def integer():
    return first(many1(digit())).map(lambda result: Constant(int(''.join(result))))

def digit():
    return satisfy(lambda c : c.isdigit())

def identifier():
    return first(many1(alpha())).map(lambda result: Variable(''.join(result)))

def alpha():
    return satisfy(lambda c : c.isalpha())

class Expression:
    def __add__(self, other):
        return Addition(self, other)
    
    def __sub__(self, other):
        return Subtraction(self, other)

    def __mul__(self, other):
        return Multiplication(self, other)

    def __truediv__(self, other):
        return Division(self, other)


class Constant(Expression):
    def __init__(self, value):
        self.value = value
    
    def __eq__(self, other):
        return isinstance(other, Constant) and self.value == other.value
    
    def __repr__(self):
        return f'Constant({self.value})'

class Variable(Expression):
    def __init__(self, name):
        self.name = name

    def __eq__(self, other):
        return isinstance(other, Variable) and self.name == other.name
    
    def __repr__(self):
        return f'Variable(\'{self.name}\')'

class Addition(Expression):
    def __init__(self, left, right):
        self.left = left
        self.right = right
 
    def __eq__(self, other):
        return isinstance(other, Addition) and self.left == other.left and self.right == other.right
    
    def __repr__(self):
        return f'Addition({self.left}, {self.right})'

class Subtraction(Expression):
    def __init__(self, left, right):
        self.left = left
        self.right = right
 
    def __eq__(self, other):
        return isinstance(other, Subtraction) and self.left == other.left and self.right == other.right
    
    def __repr__(self):
        return f'Subtraction({self.left}, {self.right})'

class Multiplication(Expression):
    def __init__(self, left, right):
        self.left = left
        self.right = right
 
    def __eq__(self, other):
        return isinstance(other, Multiplication) and self.left == other.left and self.right == other.right
    
    def __repr__(self):
        return f'Multiplication({self.left}, {self.right})'

class Division(Expression):
    def __init__(self, left, right):
        self.left = left
        self.right = right
 
    def __eq__(self, other):
        return isinstance(other, Division) and self.left == other.left and self.right == other.right
    
    def __repr__(self):
        return f'Division({self.left}, {self.right})'

if __name__ == '__main__':
    assert Constant(37) == Constant(37)
    assert Variable('foo') == Variable('foo')
    assert (Constant(37) + Constant(51)) == Addition(Constant(37), Constant(51))
    assert (Variable('foo') - Constant(37)) == Subtraction(Variable('foo'), Constant(37))
    assert (Constant(37) * Constant(51)) == Multiplication(Constant(37), Constant(51))
    assert (Variable('foo') / Constant(37)) == Division(Variable('foo'), Constant(37))

    assert integer()('37') == [(Constant(37), '')]
    assert identifier()('foo') == [(Variable('foo'), '')]

    assert factor()('37') == [(Constant(37), '')]
    assert factor()('foo') == [(Variable('foo'), '')]

    assert term()('37') == [(Constant(37), '')]
    assert term()('37*foo') == [(Multiplication(Constant(37), Variable('foo')), '')]
    assert term()('bar/51') == [(Division(Variable('bar'), Constant(51)), '')]

    assert expression()('37+foo*51') == [(Addition(Constant(37), Multiplication(Variable('foo'), Constant(51))), '')]

File: workshop/example/node/expression.js
const assert = require('assert').strict
const Parser = require('./parser')
const Util = require('./util')

class Expression {
    plus(other) {
        return new Addition(this, other)
    }

    minus(other) {
        return new Subtraction(this, other)
    }

    times(other) {
        return new Multiplication(this, other)
    }

    divideBy(other) {
        return new Division(this, other)
    }
}


class Constant extends Expression {
    constructor(value) {
        super()
        this.value = value
    }

    equals(other) {
        return other instanceof Constant && this.value === other.value
    }

    representation() {
        return 'Constant(' + this.value + ')'
    }
}

class Variable extends Expression {
    constructor(name) {
        super()
        this.name = name
    }

    equals(other) {
        return other instanceof Variable && this.name === other.name
    }

    representation() {
        return 'Variable(' + this.name + ')'
    }
}

class Addition extends Expression {
    constructor(left, right) {
        super()
        this.left = left
        this.right = right
    }

    equals(other) {
        return other instanceof Addition &&
            this.left.equals(other.left) &&
            this.right.equals(other.right)
    }

    representation() {
        return 'Addition(' + this.left + ',' + this.right + ')'
    }
}

class Subtraction extends Expression {
    constructor(left, right) {
        super()
        this.left = left
        this.right = right
    }

    equals(other) {
        return other instanceof Subtraction &&
            this.left.equals(other.left) &&
            this.right.equals(other.right)
    }

    representation() {
        return 'Subtraction(' + this.left + ',' + this.right + ')'
    }
}

class Multiplication extends Expression {
    constructor(left, right) {
        super()
        this.left = left
        this.right = right
    }

    equals(other) {
        return other instanceof Multiplication &&
            this.left.equals(other.left) &&
            this.right.equals(other.right)
    }

    representation() {
        return 'Multiplication(' + this.left + ',' + this.right + ')'
    }
}

class Division extends Expression {
    constructor(left, right) {
        super()
        this.left = left
        this.right = right
    }

    equals(other) {
        return other instanceof Division &&
            this.left.equals(other.left) &&
            this.right.equals(other.right)
    }

    representation() {
        return 'Division(' + this.left + ',' + this.right + ')'
    }
}

function expression() {
    return Util.first(Util.oneOf([
        Parser.map(Util.consecutive([
            term(),
            expressionOperator(),
            Parser.lazy(expression)
        ]), function (result) { return result[1](result[0], result[2]) }),
        term()
    ]))
}

function expressionOperator() {
    return Util.oneOf([
        Parser.map(Parser.character('+'), function () { return function (l, r) { return new Addition(l, r) } }),
        Parser.map(Parser.character('-'), function () { return function (l, r) { return new Subtraction(l, r) } }),
    ])
}


function term() {
    return Util.first(Util.oneOf([
        Parser.map(Util.consecutive([
            factor(),
            termOperator(),
            lazy(term)
        ]), function (result) { return result[1](result[0], result[2]) }),
        factor()
    ]))
}

function termOperator() {
    return Util.oneOf([
        Parser.map(Parser.character('*'), function () { return function (l, r) { return new Multiplication(l, r) } }),
        Parser.map(Parser.character('/'), function () { return function (l, r) { return new Division(l, r) } }),
    ])
}

function factor() {
    return Util.oneOf([
        integer(),
        identifier(),
        Util.parenthesized(Parser.lazy(factor))
    ])
}

function integer() {
    return Parser.map(Util.greedy1(digit()), function (result) {
        return new Constant(parseInt(result.join('')))
    })
}

function digit() {
    return Parser.satisfy(function (c) { return '0' <= c && c <= '9' })
}

function identifier() {
    return Parser.map(Util.greedy1(alpha()), function (result) {
        return new Variable(result.join(''))
    })
}

function alpha() {
    return Parser.satisfy(function (c) { return ('a' <= c && c <= 'z') || ('A' <= c && c <= 'Z') })
}

assert.ok((new Constant(37)).equals(new Constant(37)))
assert.ok((new Variable('foo')).equals(new Variable('foo')))
assert.ok((new Constant(37).plus(new Constant(51))).equals(new Addition(new Constant(37), new Constant(51))))
assert.ok((new Variable('foo').minus(new Constant(37))).equals(new Subtraction(new Variable('foo'), new Constant(37))))
assert.ok((new Constant(37).times(new Constant(51))).equals(new Multiplication(new Constant(37), new Constant(51))))
assert.ok((new Variable('foo').divideBy(new Constant(37))).equals(new Division(new Variable('foo'), new Constant(37))))

function test(parser, input, expectation) {
    result = parser(input)
    assert.equal(result[0][1], '')
    assert.ok(expectation.equals(result[0][0]))
}

test(integer(), '37', new Constant(37))
test(identifier(), 'foo', new Variable('foo'))

test(factor(), '37', new Constant(37))
test(factor(), 'foo', new Variable('foo'))
test(factor(), '(37)', new Constant(37))

test(term(), '37', new Constant(37))
test(term(), '37*foo', new Multiplication(new Constant(37), new Variable('foo')))
test(term(), 'bar/51', new Division(new Variable('bar'), new Constant(51)))

test(expression(), '37+foo*51', new Addition(new Constant(37), new Multiplication(new Variable('foo'), new Constant(51))))